### PR TITLE
Draggable annotations

### PR DESCRIPTION
**Pull request by LukasPaczos**
Closes https://github.com/mapbox/mapbox-plugins-android/issues/703.

ezgif com-video-to-gif 14

### CODE REVIEW

**Comment by LukasPaczos** [comment]
This one is ready for a review.

ezgif com-video-to-gif 15

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java [hunk 1]
```diff
@@ -265,16 +295,26 @@ public void onMapLongClick(@NonNull LatLng point) {
...
+  }
+
+  @Nullable
+  T queryMapForFeatures(@NonNull PointF point) {
+    List<Feature> features = mapboxMap.queryRenderedFeatures(point, getAnnotationLayerId());
+    if (!features.isEmpty()) {
+      long id = features.get(0).getProperty(getAnnotationIdKey()).getAsLong();
+      return annotations.get(id);
     }
+    return null;
```
Comment (tobrun) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java]: there seems to be some shared code between these 2 methods.

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Circle.java [hunk 2]
```diff
@@ -35,12 +42,23 @@
...
   }
 
+  /**
+   * Get the LatLng of the circle, which represents the location of the circle on the map
+   *
+   * @return the location of the circle
+   */
+  @NonNull
+  public LatLng getLatLng() {
+    Point point = (Point) geometry;
```
Comment (tobrun) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Circle.java]: not related to this PR, but something that comes to mind now, do we need type geometry with generics? (avoids casting in this case).

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java [hunk 3]
```diff
@@ -49,6 +56,29 @@ public void setLatLngs(List<List<LatLng>> latLngs) {
...
+      for (List<Point> innerPoints : coordinates) {
+        List<LatLng> innerList = new ArrayList<>();
+        for (Point point : innerPoints) {
+          innerList.add(new LatLng(point.latitude(), point.longitude()));
+        }
+        latLngs.add(innerList);
+      }
+      return latLngs;
+    }
+    return null;
```
Comment (tobrun) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java]: do we want nullability for this? Returning an empty list would suffice and would make the method `@NonNull`

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java [hunk 4]
```diff
@@ -148,4 +187,33 @@ public String getFillPattern() {
...
+          }
+          innerPoints.add(Point.fromLngLat(latLng.getLongitude(), latLng.getLatitude()));
+        }
+        resultingPoints.add(innerPoints);
+      }
+
+      return Polygon.fromLngLats(resultingPoints);
+    }
+
+    return null;
```
Comment (tobrun) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java]: do we want nullability for this? Returning an empty list would suffice and would make the method `@NonNull`

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java (hunk 3 above)
Comment (LukasPaczos) [reply plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java]: Good point

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java (hunk 4 above)
Comment (LukasPaczos) [reply plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Fill.java]: In this case, we are returning `null` to indicate that the resulting geometry is invalid and we should not set it, leaving the old one being displayed

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java (hunk 1 above)
Comment (LukasPaczos) [reply plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java]: yup, going to fix this up

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Circle.java (hunk 2 above)
Comment (LukasPaczos) [reply plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/Circle.java]: we can possibly do it to clean up those casts :+1:

**Pull request merged**